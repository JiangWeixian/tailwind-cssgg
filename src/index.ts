import plugin from 'tailwindcss/plugin'
import { buildCssggIcon } from './lib'

export const cssgg = plugin((props) => {
  const components = buildCssggIcon()
  props.matchComponents(
    {
      gg: (value) => {
        if (typeof value === 'string') {
          console.warn(
            'Unable able to find vaild css.gg icon, pelase visit https://css.gg/ for more information about available icon names, e.g. `.gg-add-r`',
          )
          return {}
        }
        return value
      },
    },
    { values: components },
  )
})
