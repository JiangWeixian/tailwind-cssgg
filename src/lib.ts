import cssggJson from 'css.gg/icons/all.json'

export const generateIconComponent = (key: string, css: string) => {
  const resolvedKey = new RegExp(`.gg-${key}`, 'g')
  const styleObject: Record<string, Record<string, string>> = {}

  css.split('}').forEach((rule) => {
    if (rule.trim() === '') return

    const [selector, styles] = rule.split('{')
    const cleanSelector = selector.trim().replace('"', '')
    const cleanStyles = {}
    if (!styles) {
      return
    }
    styles?.replace(/([^:]+):*([^;]+);*/gm, (_, prop, value) => {
      cleanStyles[prop] = value
      return ''
    })

    styleObject[cleanSelector.replace(resolvedKey, '&')] = cleanStyles
  })
  return styleObject
}

export const buildCssggIcon = () => {
  const cssggIcons = {}
  Object.keys(cssggJson).forEach((key) => {
    const data = cssggJson[key][0][0]
    cssggIcons[key] = generateIconComponent(key, data.css)
  })
  return cssggIcons
}
