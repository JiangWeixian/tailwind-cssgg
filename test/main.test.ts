import { test, expect } from 'vitest'
import cssggJson from 'css.gg/icons/all.json'
import { generateIconComponent, buildCssggIcon } from '../src/lib'

test('icon components styles', () => {
  const collections = buildCssggIcon()
  expect(collections['add-r']).toMatchSnapshot()
})

test('generate icon component', () => {
  const raw = generateIconComponent('zeit', cssggJson.zeit[0][0].css)
  expect(raw).toMatchSnapshot()
})
