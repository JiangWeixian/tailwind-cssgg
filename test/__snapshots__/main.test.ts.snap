// Vitest Snapshot v1

exports[`generate icon component 1`] = `
{
  "&": {
    "border-bottom": "18px solid currentColor",
    "border-left": "9px solid transparent",
    "border-right": "9px solid transparent",
    "box-sizing": "border-box",
    "display": "block",
    "height": "0px",
    "position": "relative",
    "transform": "scale(var(--ggs,1))",
    "width": "0px",
  },
}
`;

exports[`icon components styles 1`] = `
{
  "&": {
    "border": "2px solid",
    "border-radius": "4px",
    "box-sizing": "border-box",
    "display": "block",
    "height": "22px",
    "position": "relative",
    "transform": "scale(var(--ggs,1))",
    "width": "22px",
  },
  "&::after": {
    "height": "10px",
    "left": "8px",
    "top": "4px",
    "width": "2px",
  },
  "&::after,.gg-add-r::before": {
    "background": "currentColor",
    "border-radius": "5px",
    "box-sizing": "border-box",
    "content": "\\"\\"",
    "display": "block",
    "height": "2px",
    "left": "4px",
    "position": "absolute",
    "top": "8px",
    "width": "10px",
  },
}
`;
